omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java

```diff
@@ -143,26 +152,46 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
 				try {
 					Context.addProxyPrivilege(PrivilegeConstants.GET_USERS);
 					user = Context.getUserService().getUserByUsername(username);
-				}
-				finally {
+				} finally {
 					Context.removeProxyPrivilege(PrivilegeConstants.GET_USERS);
 				}
 				
+				String secretQuestion = null;
+				if (user != null) {
+					secretQuestion = Context.getUserService().getSecretQuestion(user);
+				}
+				
 				// check the secret question again in case the user got here "illegally"
-				if (user == null /*|| StringUtils.isEmpty(user.getSecretQuestion())*/) {
+				if (user == null) {
+					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
+					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
+					request.setAttribute("secretQuestion", getRandomFakeSecretQuestion(username));
+				} else if (secretQuestion == null || secretQuestion.equals("")) {
 					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.question.empty");
-				} else if (/*user.getSecretQuestion() != null &&*/ Context.getUserService().isSecretAnswer(user, secretAnswer)) {
+				} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
 					
 					StringBuilder randomPassword = new StringBuilder();
-					for (int i = 0; i < 8; i++) {
-						randomPassword.append(String.valueOf((Math.random() * (127 - 48) + 48)));
+					ArrayList<Character> buildRandomPassword = new ArrayList<Character>();
+					
+					//Ensure atleast 1 uppercase, 1 lowercase as well as 1 digit in password (Otherwise, exception is thrown)
+					buildRandomPassword.add((char) (65 + (int) Math.abs(Math.random()) % 26));
```

Comment:
I added a code snippet of what you need to use to create a valid and random password on one of the pull requests you closed. You need to use a combination of RandomStringUtils.alphaNumeric and RandomStringUtils.random and convert the returned letter to upper case and pre-append it to the alpha numeric string , the length would need to depend on the value of the global property for password minimum length